src/AWS/Signers/V4.elm

```diff
@@ -73,31 +75,45 @@ addAuthorization :
     AWS.Config.Service
     -> AWS.Config.Credentials
     -> Date
+    -> UnsignedRequest a
     -> List ( String, String )
     -> List ( String, String )
-addAuthorization config creds date headers =
+addAuthorization config creds date req headers =
     [ ( "X-Amz-Date", formatDate date )
-    , ( "Authorization", authorization creds date config headers )
+    , ( "Authorization", authorization creds date config req headers )
     ]
         |> List.append headers
 
 
-authorization : Credentials -> Date -> AWS.Config.Service -> List ( String, String ) -> String
-authorization creds date config headers =
-    [ "AWS4-HMAC-SHA256 Credentials="
-        ++ creds.accessKeyId
-        ++ "/"
-        ++ credentialsString date config
-    , "SignedHeaders="
-        ++ signedHeaders headers
-    , "Signature="
-        ++ signature creds
-    ]
-        |> String.join ", "
-
-
-credentialsString : Date -> AWS.Config.Service -> String
-credentialsString date config =
+authorization :
+    Credentials
+    -> Date
+    -> AWS.Config.Service
+    -> UnsignedRequest a
+    -> List ( String, String )
+    -> String
+authorization creds date config req headers =
+    let
+        canon =
+            canonical req.method req.path headers req.params
+
+        scope =
+            credentialScope date config
+    in
+        [ "AWS4-HMAC-SHA256 Credential="
+            ++ creds.accessKeyId
+            ++ "/"
+            ++ scope
+        , "SignedHeaders="
+            ++ signedHeaders headers
+        , "Signature="
+            ++ signature creds config date (stringToSign algorithm date scope canon)
```

Comment:
I'm not terribly happy with the way these two functions are defined. Open to suggestions on improving the data sharing and API.